### DISCUSSION TITLE
eQSL in Cron?

### DISCUSSION THREAD

**Discussion by makeitrad** [magicbug/Cloudlog#751] [Ideas / Feature Requests]
Is there a way to put the uploading of eQLS contacts in the cron similar to LOTW or the QRZ Logbook?
Thanks & 73.
Brad
N1KN

**Comment by magicbug** [comment]
Currently not, although just requires a few code changes, I'll try get it added.

**Comment by makeitrad** [comment]
Thanks Peter! With that I'd have little to no reason to go into the admin panel :)
Much appreciate the response on the weekend. 
Happy Holidays,
Brad

**Comment by magicbug** [comment]
Completely agree its the only time I go in the admin area usually.

**Comment by m0urs** [comment]
Is this still something which we would get in near future?

**Reply by m0urs** [reply]
Just would like to push this issue up a bit. That is really something which I am still missing as I do need to do that manually (and often forget to do ...) :-)

**Comment by AndreasK79** [comment]
@m0urs merged into dev. Call it with cloudlog-url/index.php/eqsl/upload

**Reply by m0urs** [reply]
Hm, it seems that this URL does still not do anything?
And is the download option planned as well?

**Reply by phl0** [reply]
Yeah it shows no output. But upload should work nonetheless.

**Reply by m0urs** [reply]
Unfortunately it does not upload anything here :-(

**Reply by phl0** [reply]
Any error messages in the log? It worked on my tests.

**Reply by m0urs** [reply]
Is there a special Cloudlog log? In the normal Apache error logs there is no message. Just to make sure, this is the command I have tried to use from the command line. Before I added one test QSO in Cloudlog which is visible if I am using the eQSL upload view.

curl  https://cloudlog.urspringer.de/index.php/eqsl/upload

**Reply by m0urs** [reply]
This is the log entry in the Apache access log:

82.165.xxx.yyy - - [05/May/2023:15:23:07 +0200] "GET /index.php/eqsl/upload HTTP/2.0" 302 1027 "-" "curl/7.64.0" - cloudlog.urspringer.de:443

There is a 302 message which means "redirect" which seems strange to me? Especially as there is no other request afterwards.

**Reply by AndreasK79** [reply]
What is your log level set at? 

config.php: $config['log_threshold'] = 4;

You should then find a log in application/logs

**Reply by m0urs** [reply]
Just set "$config['log_threshold'] = 4;" but there is no log appearing. Neither while running curl from the command line nor while playing with Cloudlog via the WebUI?

I also, for testing, disabled the index.php suppression which I normally have activated, but this did not change anything.

**Reply by m0urs** [reply]
The issue with the log file was a access rights problem. Apache user had no access to write into that directory. This is the content of the PHP file created now while running the curl command:

```
INFO - 2023-05-05 13:40:28 --> Config Class Initialized
INFO - 2023-05-05 13:40:28 --> Hooks Class Initialized
INFO - 2023-05-05 13:40:28 --> UTF-8 Support Enabled
INFO - 2023-05-05 13:40:28 --> Utf8 Class Initialized
INFO - 2023-05-05 13:40:28 --> URI Class Initialized
INFO - 2023-05-05 13:40:28 --> Router Class Initialized
INFO - 2023-05-05 13:40:28 --> Output Class Initialized
INFO - 2023-05-05 13:40:28 --> Security Class Initialized
INFO - 2023-05-05 13:40:28 --> Input Class Initialized
INFO - 2023-05-05 13:40:28 --> Language Class Initialized
INFO - 2023-05-05 13:40:28 --> Loader Class Initialized
INFO - 2023-05-05 13:40:28 --> Config file loaded: /var/www/cloudlog.urspringer.de/html/application/config/cloudlog.php
INFO - 2023-05-05 13:40:28 --> Config file loaded: /var/www/cloudlog.urspringer.de/html/application/config/bands.php
INFO - 2023-05-05 13:40:28 --> Config file loaded: /var/www/cloudlog.urspringer.de/html/application/config/lotw.php
INFO - 2023-05-05 13:40:28 --> Helper loaded: url_helper
INFO - 2023-05-05 13:40:28 --> Helper loaded: security_helper
INFO - 2023-05-05 13:40:28 --> Language file loaded: language/english/general_words_lang.php
INFO - 2023-05-05 13:40:28 --> Database Driver Class Initialized
INFO - 2023-05-05 13:40:28 --> Session: Class initialized using 'files' driver.
DEBUG - 2023-05-05 13:40:28 --> cURL Class Initialized
INFO - 2023-05-05 13:40:28 --> Migrations Class Initialized
INFO - 2023-05-05 13:40:28 --> Language file loaded: language/english/migration_lang.php
INFO - 2023-05-05 13:40:28 --> Database Forge Class Initialized
INFO - 2023-05-05 13:40:28 --> Model "Options_model" initialized
INFO - 2023-05-05 13:40:28 --> Controller Class Initialized
INFO - 2023-05-05 13:40:28 --> Helper loaded: form_helper
INFO - 2023-05-05 13:40:28 --> Model "User_Model" initialized

```

**Reply by m0urs** [reply]
It does work if I am using this URL from within a browser session where I am logged in into Cloudlog. So I guess it is a a problem because you are not yet authorized. This would explain the 302 message as well as Cloudlog seems to redirect you to the login page.

**Reply by AndreasK79** [reply]
@m0urs great. I found the reason. As I am not using eQSL, I can't test this myself, so it's hard to debug remote.

**Reply by AndreasK79** [reply]
@m0urs pull dev and test. Pushed fix now :)

**Reply by m0urs** [reply]
Hm, seems that there is no change. If you are using curl with "-L" so that it follows the redirect, you can see that the login page will be called. So it is still the problem that we do not have any session authentication.

**Reply by AndreasK79** [reply]
Thanks. I removed the one place I could find where it used userid from session. I'll dig some more.

**Reply by AndreasK79** [reply]
@m0urs pushed another fix. It checked in the controller constructor if user was logged in. 

I do not get the redirect anymore.

**Reply by m0urs** [reply]
Yes, I can confirm that it is now uploading correctly! Thank you.
So, could we get download of eQSL for CRON as well? ;-)

**Reply by AndreasK79** [reply]
> So, could we get download of eQSL for CRON as well? ;-)

Yes. It is available here: https://github.com/AndreasK79/Cloudlog/tree/eqsl_cron_job

It is untested, so no PR yet. Call it with eqsl/sync

I merged upload and download into sync instead.

**Reply by m0urs** [reply]
Not sure how to pull these changes?

![image](https://user-images.githubusercontent.com/2588362/236645439-b91d5994-db23-462b-bd2f-f5c43c99e50c.png)

**Reply by AndreasK79** [reply]
I made it into a draft PR: https://github.com/magicbug/Cloudlog/pull/2112

**Reply by m0urs** [reply]
Thank you. Seems to work for me. Although I need some more real testing with new QSOs. What I have seen is that I cannot see the eQSL images anymore. However, after some investigation,it seems to be a problem which is not related to this. My last eQSL image is from April 8th :-(  Currently not sure what the reason for that could be.

**Reply by AndreasK79** [reply]
Thanks for testing. I also made so that the download outputs some info. 

Aside from that, the cron code is using the same code as the gui inside Cloudlog.

**Reply by m0urs** [reply]
I just checked a bit deeper and I think the eQSL image download is broken since a long time already or I did not understand when images are downloaded. I will oben a new issue for that. So again, I think the sync itself via command line is working as expected.

**Reply by AndreasK79** [reply]
This issue is fixed, and cron job is merged into dev.
